### PR TITLE
Frontend support for delegation and rejection

### PR DESCRIPTION
**Pull request by li-boxuan**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

1. If an agent rejects a task, the web GUI would "crash" because we don't support it on frontend yet
2. Agent delegation always shows a weird empty box when it finishes, because frontend doesn't handle agent delegate observation

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

1. Add support for rejection action on frontend
2. Show users the reason for rejection
3. Get rid of weird empty box after delegation

**Other references**

Demo (this PR, using `ManagerAgent`, with some prompt tweaks, which will be in a separate PR):

### Delegate & Rejection

image

### Delegate & Success

image

You might see some messages are seemingly redundant and not very user-friendly. This is left for https://github.com/OpenDevin/OpenDevin/issues/2309 and we may need some UI redesign. This PR makes it at least work end-to-end.

### CODE REVIEW

**Comment by li-boxuan** [comment]
Asking CodeActAgent to use Google search to find out the current president of USA:

image

The first weird `All done! What's next on the agenda?` message is due to delegated BrowsingAgent finishing its task. Seems weird at the moment. We should collapse delegate agent's activities as suggested here: https://github.com/OpenDevin/OpenDevin/issues/2309#issuecomment-2154312838

**Review by amanape — COMMENTED** [review]
LGTM!

**Comment by enyst** [comment]
The message with JSON in the last screenshot is the delegate observation, right? I think we can tweak its `message` property to display user-friendly text instead of JSON, so that it displays more naturally in the chat box.

I'm not sure exactly how, it seems it can respond with different dict keys, and currently, that property is `` by default... but maybe we can find a way to make it into a phrase that includes the relevant key values when they exist?

**Comment by li-boxuan** [comment]
> I'm not sure exactly, it seems it can respond with different dict keys

You are right! It could be anything. Every micro-agent, for example, has its own schema defined in `agent.yaml`.

Why don't we simply make every agent return a natural language output as `CodeActAgent`? The reason is, some agents (especially micro-agents) are designed to do a very narrow task, and they may or may not want to answer in natural language. A SQL agent, for (imaginary) example, could return SQL execution results in any pre-defined format.

What we should do here is unclear to me. I don't know how we could design delegation better and more elegant.

**Comment by enyst** [comment]
Oh, I see, I was actually thinking we might be able to retrieve this info _a little_ better... like, from the agent who should know best what it's doing or at least from the parent who asked for something, one would think it can read a response...in some way we can use, but no... not quite. 😅 

Okay, there is someone who knows our answer, but that someone is not in our code, it's... the LLM. 😅 

Then I suppose the fallback option is the LLM, right: a summary sentence. I thought delegate summaries made a lot of sense and save tokens and confusion, but it seems to me that they're simply necessary here. It would tell us what the delegate concluded or how it concluded.

For this PR, what do you think about something like: 
`Delegate agent finishes task with content: Based on search results, the current president of the USA is Joe Biden.`
`Delegate agent finishes task with answer: 0.62....`
`Delegate agent finishes task with reason: Not a valid git repository....`

In other words, I'm wondering if the best we can do for now is just the absence of too obvious formatting, since it's not a code block...

**Comment by li-boxuan** [comment]
> Then I suppose the fallback option is the LLM, right: a summary sentence. I thought delegate summaries made a lot of sense and save tokens and confusion, but it seems to me that they're simply necessary here. It would tell us what the delegate concluded or how it concluded.

Exactly. Your work would be very important here.

**Review by enyst — APPROVED** [review]

**Comment by li-boxuan** [comment]
image

@enyst  Applied your suggestion. Looks better now, thanks!

**Comment by tobitege** [comment]
in the image, I'd find it confusing, why suddenly a "**Delegate agent...**" is mentioned, without first seeing something like "Task has been delegated..." in chat?
Just as an observation.

**Comment by tobitege** [comment]
Let me rephrase my observation:
from a _user_ perspective, the 2nd and 3rd agent messages are not relevant/confusing.
Even though they may be _technically_ needed/wanted in the backend.

**Comment by li-boxuan** [comment]
> Let me rephrase my observation: from a _user_ perspective, the 2nd and 3rd agent messages are not relevant/confusing. Even though they may be _technically_ needed/wanted in the backend.

The 1st agent message will be auto-collapsed after #2309

The 2nd agent message is a summary of what has been done by sub-agent

The 3rd agent message is specific to rejection (not really delegate-related). The "the agent" here refers to the current agent (at that time), which is the top-most agent.

**Comment by tobitege** [comment]
> > Let me rephrase my observation: from a _user_ perspective, the 2nd and 3rd agent messages are not relevant/confusing. Even though they may be _technically_ needed/wanted in the backend.
> 
> The 1st agent message will be auto-collapsed after #2309
> 
> The 2nd agent message is a summary of what has been done by sub-agent
> 
> The 3rd agent message is specific to rejection (not really delegate-related). The "the agent" here refers to the current agent (at that time), which is the top-most agent.

Right. That is all transparent and obvious to you, that it is a sub agent.
It's probably just me, but with my "user hat" on, I don't have a need to know, other than the result, i.e rejection happened for a reason.
But I'll leave it be now, please continue. 😁

**Comment by li-boxuan** [comment]
> without first seeing something like "Task has been delegated..." in chat?

That's a good point. I'll add support for that.

**Comment by li-boxuan** [comment]
I have to admit the UX for delegates & rejection are independently awful 😆

**Comment by li-boxuan** [comment]
@tobitege Applied your suggestion

> without first seeing something like "Task has been delegated..." in chat?

image

**Comment by li-boxuan** [comment]
We still have some low-hanging fruits that can improve UX by a bit: e.g. don't say "delegate" agent, say what exactly agent it is.

i.e.

> I am asking CommitWriterAgent to...

sounds better than

> I am asking Delegate Agent to...

Similarly, we shouldn't say

> Task is rejected by the agent

we should say

> Task is rejected by CommitWriterAgent

I'd leave it as good first issue.
